### PR TITLE
Disable ruby-lsp-rspec's definition listener outside test files

### PR DESCRIPTION
**Pull request by MaicolBen**
Here is what happens when I add the addon
Screenshot 2024-10-16 at 11 15 51

### CODE REVIEW

File: lib/ruby_lsp/ruby_lsp_rspec/definition.rb [hunk 1]
```diff
@@ -42,6 +42,7 @@ def on_call_node_enter(node)
         entries.each do |entry|
           # Technically, let can be defined in a different file, but we're not going to handle that case yet
           next unless entry.file_path == @uri.to_standardized_path
+          next unless entry.file_path.end_with?('spec.rb')
```
Comment (MaicolBen) [inline lib/ruby_lsp/ruby_lsp_rspec/definition.rb:45]: This breaks tests that use tempfiles, do you know how can I make tempfiles use `spec.rb` files? Or do you know any other way to check it's an rspec. I have seen used `node.receiver && node.receiver&.slice != "RSpec"` but it doesn't work always.

**Comment by st0012** [comment]
I roughly understand the issue but it can not be caused by the listener you're trying to patch because it only returns definitions that are also from the same file ([condition](https://github.com/st0012/ruby-lsp-rspec/blob/main/lib/ruby_lsp/ruby_lsp_rspec/definition.rb#L44)).
This means the duplicated definition you see is returned by `ruby-lsp`'s own definition listener. And since it only happens when you install `ruby-lsp-rspec`, I guess it's because you have a `let(:sync_with_intercom)` or `subject(:sync_with_intercom)` in your spec files?

**Comment by MaicolBen** [comment]
Thank you for replying so quickly! 

> I roughly understand the issue but it can not be caused by the listener you're trying to patch because it only returns definitions that are also from the same file ([condition](https://github.com/st0012/ruby-lsp-rspec/blob/main/lib/ruby_lsp/ruby_lsp_rspec/definition.rb#L44)).

Yes, so both core and LSP rspec return definitions

> I guess it's because you have a let(:sync_with_intercom) or subject(:sync_with_intercom) in your spec files?

That doesn't exist at all, this happens with every definition of every file (not if the definition is in another file). Also, how can you explain that if I remove my new check, the test returns 2 definitions instead of 1? 

I'm curious about your setup, since I cannot get working this addon without including in the gemfile.

**Comment by st0012** [comment]
It took me a while to find a reproduceable case in my projects but I can confirm that it is the listener's fault 👍 
But I think we can simply not initializing this listener in non-spec files. So the solution would be more like this:

```diff
diff --git a/lib/ruby_lsp/ruby_lsp_rspec/addon.rb b/lib/ruby_lsp/ruby_lsp_rspec/addon.rb
index 98d13ff..3cde027 100644
--- a/lib/ruby_lsp/ruby_lsp_rspec/addon.rb
+++ b/lib/ruby_lsp/ruby_lsp_rspec/addon.rb
@@ -64,6 +64,8 @@ module RubyLsp
         ).void
       end
       def create_definition_listener(response_builder, uri, node_context, dispatcher)
+        return unless uri.to_standardized_path&.end_with?("_test.rb") || uri.to_standardized_path&.end_with?("_spec.rb")
+
         Definition.new(response_builder, uri, node_context, T.must(@index), dispatcher)
       end
```

**Comment by MaicolBen** [comment]
Nice! Thank you, it's because I don't know much how the LSP addon system works. This solution has the same issue with specs [than mine](https://github.com/st0012/ruby-lsp-rspec/pull/46#discussion_r1807051883), any ideas?

**Comment by st0012** [comment]
On my side the test failed without the change and passed when the above diff is applied. Can you first push the lib code and we can see what's failing on CI?

**Comment by st0012** [comment]
Ah ok I get what you mean not. It's other tests that are failing.

You can do this to create tempfiles that end with specific postfix: `Tempfile.new(["", "_fake_spec.rb"])`

**Comment by MaicolBen** [comment]
Nice! Thank you!

**Review by st0012 — APPROVED** [review]
Thank you so much for reporting the issue and the PR 😄

**Comment by st0012** [comment]
I've cut a release for this: https://github.com/st0012/ruby-lsp-rspec/releases/tag/v0.1.17

**Pull request merged**